### ISSUE TITLE
Allow for localizers to vary the number of elements in nested DOM

### ISSUE THREAD

**Issue by zbraniecki** [zbraniecki/fluent-domoverlays-js#2]
[netError.dtd](https://searchfox.org/mozilla-central/source/browser/locales/en-US/chrome/overrides/netError.dtd) case requires us to handle a scenario where the users may want to vary the number of `<li/>` elements in two ways.

First, the structure of the DOM between two variants of the message differs:

```
network-down=
  <ul>
    <li>The network is down.</li>
    <li>That's sad.</li>
  </ul>

network-up=
  <ul>
    <li>The network is up.</li>
    <li>That's nice.</li>
    <li>Let's go suring!</li>
  </ul>
```

We can either handle the same source structure with ability to translate it via one of the two messages, or we need to ask developers to provide two source structures and flip their visibility.

Second, there's a potential value in allowing localizers to customize the number of `<li/>` elements within the message by, for example, splitting a single `<li/>` into multiple or removing a `<li/>`.

That would require us to change how we whitelist allowed and expected elements since the prototype I put forth currently requires the source structure to be maintained in the translation which prevents localizers from modifying it.

**Comment by flodolo** [comment]
> Second, there's a potential value in allowing localizers to customize the number of `<li/>` elements within the message by, for example, splitting a single `<li/>` into multiple or removing a `<li/>`.

For the record, I don't think this is a realistic case, at least for Firefox. Each bullet point is a well contained element (action), and they should really not be split.

**Comment by zbraniecki** [comment]
The challenge with this proposal is that it is not compatible with Item 7 (retain elements between retranslations).

In particular, what might happen is:

```html
<ul id="ul1" data-l10n-id="key1">
</ul>
```

translation1: 
```properties
key1 =
  <li>Item 1</li>
  <li>Item 2</li>
  <li>Item 3</li>
```

translation2:
```properties
key1 =
  <li>Item 1</li>
  <li>Item 2</li>
```

```js
let ul1 = document.getElementById("ul1");
await document.l10n.translateFragment(ul1); // into translation1
let li3 = ul1.getElementsByTagName('li')[2]; // li:nth(3)
await document.l10n.translateFragment(ul1); // into translation2
```
this leaves `li3` element in memory but detached from the DOM. I believe that would be a big papercut for developers.

**Comment by Pike** [comment]
I personally prefer this Item 9 over Item 7.

I can picture various use cases for `l10n.setAttributes` doing interesting things to the DOM underneath. And it just feels weird that you'd have to change a ton of code if one error message grows from 2 to 3 list items. That is, not only do you need to change the msg ID, but you also need to change the container they're hooked up to, and possibly even create that container with blank DOM content.

Questions around node identity or re-use seem secondary to me there.

**Comment by zbraniecki** [comment]
@msujaws, @gijsk?

**Comment by gijsk** [comment]
> We can either handle the same source structure with ability to translate it via one of the two messages, or we need to ask developers to provide two source structures and flip their visibility.

Not really; the source structure is very similar so it's easy to adapt it for the different messages at the same time where we make the "right" call to document.l10n.setAttributes (e.g. by removing superfluous elements for messages that require fewer). This is exactly what we're in the process of doing for about:blocked .

> Second, there's a potential value in allowing localizers to customize the number of `<li/>` elements within the message by, for example, splitting a single `<li/>` into multiple or removing a `<li/>`.

As @flod said, I don't think there is.

> ```js
> let ul1 = document.getElementById("ul1");
> await document.l10n.translateFragment(ul1); // into translation1
> let li3 = ul1.getElementsByTagName('li')[2]; // li:nth(3)
> await document.l10n.translateFragment(ul1); // into translation2
> ```
> 
> this leaves `li3` element in memory but detached from the DOM. I believe that would be a big papercut for developers.

Why? Papercut in which way? This is possible today, and as long as `li3` goes out of scope on the JS side, it will just get GC'd. For other things, they'll get cleared up when the page is unloaded, one hopes (ie script won't stay live, or if it does, will be forced to null out all its references to avoid leaks irrespective of this particular point).

> Questions around node identity or re-use seem secondary to me there.

But they're not... if I have:

```html
<p data-l10n-id="foo"><a id="link"></p>
```

```
foo = "Blah blah <a>Click here</a>"
```

```js
document.getElementById("link").href = "blah";
```

then I don't want retranslation to clobber my href, also because there's no events for when translation happens, so nothing I can do about it as a developer to avoid things going missing.

**Comment by Pike** [comment]
https://phabricator.services.mozilla.com/D23611 is an interesting use-case for this bug. Might be useful to look at it's final version, the one that's there right now will still see heavy changes, I suppose.

To clarify on my comment on node identity. Keeping developer-facing artifacts like `a.href` is the job of `data-l10n-name` in my book. Right now, the overlay code creates a new node with the same attributes, so, as far as attributes go, the node looks the same. But there's no identity. My comment was in particular about un-named nodes.

**Comment by gijsk** [comment]
> To clarify on my comment on node identity. Keeping developer-facing artifacts like `a.href` is the job of `data-l10n-name` in my book. Right now, the overlay code creates a new node with the same attributes, so, as far as attributes go, the node looks the same. But there's no identity. My comment was in particular about un-named nodes.

Sure, I probably shouldn't have used `href` - an `onclick` property that runs custom JS when the user clicks the link, or the equivalent `addEventListener` (which is not that uncommon), would have been a better example. Just copying attributes isn't sufficient for that. Even copying event listeners (which the event listener service in gecko might allow you to do in privileged code) might not be sufficient because the JS code will be confused about where to add/remove listeners when node identities are not kept (ie if the JS keeps a reference and tries to remove it from there instead of the newly minted node that fluent would move the listener to).

**Comment by zbraniecki** [comment]
> Not really; the source structure is very similar so it's easy to adapt it for the different messages at the same time where we make the "right" call to document.l10n.setAttributes (e.g. by removing superfluous elements for messages that require fewer).

I don't think I understand this point @gijsk. If you have two locales that have different number of child elements for the same l10n-id, then a locale change will trigger retranslation into a new list of child elements.

We then either have to allow for arbitrary number of child elements, or restrict the source and all translation to have the same length, in which case for cases where we do have a variable between, say, 2 and 3, we'll need two DOM fragments in source and two data-l10n-ids - one for 2, and one for 3.
Am I missing some other option?

> As @flod said, I don't think there is.

I find this value very small, and the complexity and limitation it brings vastly outweighing it.

> Why? Papercut in which way? This is possible today, and as long as li3 goes out of scope on the JS side, it will just get GC'd.

Which means that in all cases where we do `Structure.elements['url-link']` and cache the result of `getElement*` to that `elements` list we'll keep the element accessible and you'll even be able to operate on it but you'll likely never notice that it was removed from the DOM, and then, maybe, re-added as a new element that is now visible in your source code, but is not the same as the reference you keep in JS.

I'd like to mitigate that and that's what Item 3 and Item 7 bring. I think that's incompatible with the Item 9 proposal, since it requires ability to remove elements from source and re-add them with different identity.

> an onclick property that runs custom JS when the user clicks the link, or the equivalent addEventListener (which is not that uncommon), would have been a better example.

I find this use case important.

I now landed the avoid-churn branch and removed discussed features from master into PRs.

**Comment by zbraniecki** [comment]
I'm going to work on a PR to try to add this feature. It'd work like this:

```html
<div data-l10n-id="key1" data-l10n-whitelist="ul, ul > li">
</div>
```

```html
key1 = This is a list of errors with
  <ul>
    <li>Bullet 1</li>
    <li>Bullet 2</li>
    <li>Bullet 3</li>
  </ul>
```

In order for this to work, we'll need to break Item 3 (if an element is missing in translation and part of whitelist, I'll remove it from the source). We'll also need to lift at least some of the Item 7 (an element may come and go between retranslations and in particular the element is not in the source until translation).

@Pike - does it seem like what you're suggesting?

**Comment by gijsk** [comment]
> > Not really; the source structure is very similar so it's easy to adapt it for the different messages at the same time where we make the "right" call to document.l10n.setAttributes (e.g. by removing superfluous elements for messages that require fewer).
> 
> I don't think I understand this point @gijsk. If you have two locales that have different number of child elements for the same l10n-id, then a locale change will trigger retranslation into a new list of child elements.

I was responding to your point in the initial comment, which was about 2 messages with different IDs, (presumptively) in the same translation.

> We then either have to allow for arbitrary number of child elements, or restrict the source and all translation to have the same length, in which case for cases where we do have a variable between, say, 2 and 3, we'll need two DOM fragments in source and two data-l10n-ids - one for 2, and one for 3.
> Am I missing some other option?

As said in the other comments, I don't think a variation in non-inline/text child elements is actually realistically something we need regularly.

> > As @flod said, I don't think there is.
> 
> I find this value very small, and the complexity and limitation it brings vastly outweighing it.

So are we just violently agreeing? I mean, apparently not because you're suggesting you want a PR for this (presumably nr. 9) feature... whereas my argument is that there's no real usecase for it...

**Comment by zbraniecki** [comment]
We are violently agreeing indeed! Item 9 was added as a way to evaluate a potential need pointed out to us, and to have an opportunity to evaluate how such feature may or may not come in conflict with other items.
If it does, well need to evaluate pros and cons of each item.

I state that from my perspective the value of item 9 is low and in particular much lower than item 7 and item 3.
